### PR TITLE
Optimize hash collection

### PR DESCRIPTION
**Pull request by privat**
Some optimization when dealing with empty hash-based collection

user time for nitc nitc.nit:

* before: 0m6.640s
* after: 0m6.344s (-4.5%)

### CODE REVIEW

**Comment by xymus** [comment]
Was the `enlarge(0)` doing anything?

+1

**Comment by privat** [comment]
Yes, the enlarge(0) allocated a non-empty NativeArray. This PR remove this
mandatory allocation.

Le lun. 19 oct. 2015 16:34, Alexis Laferrière <notifications@github.com> a
écrit :

File: lib/core/collection/hash_collection.nit [hunk 1]
```diff
@@ -253,6 +257,7 @@ class HashMap[K, V]
 
 	redef fun []=(key, v)
 	do
+		if _capacity == 0 then enlarge(4)
```
Comment (Morriar) [inline lib/core/collection/hash_collection.nit]: You should comment the magic `4`.

**Comment by Morriar** [comment]
+1

**Comment by privat** [comment]
reroll: document the marig number and some more cleaning

File: lib/core/collection/hash_collection.nit [hunk 2]
```diff
@@ -442,6 +435,7 @@ class HashSet[E]
 
 	redef fun add(item)
 	do
+		if _capacity == 0 then enlarge(17) # 17 because magic in `store`
```
Comment (Morriar) [inline lib/core/collection/hash_collection.nit]: This is black magic, better back-off!

**Pull request merged**